### DISCUSSION TITLE
Need help developing in Bonsai - want to add alignment editing

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#5716] [Q&A]
@Andrej730 I'm hoping you can point me in the right direction. I am trying to learn how to develop for Bonsai so I can add infrastructure/alignment editing features.

I went through the Bonsai Demo Panel example. Using that as I guide I setup my own panel with a single button that creates a hard-coded horizontal alignment. 

The code runs fine - an alignment is created. You can see this by creating a new IFC project from the wizard (Using Ifc4x3 schema), pressing the "Build Alignment" button, then saving and re-opening the file.

My specific question is how do I update the Scene Collection and 3D view with the alignment when the "Build Alignment" button is pressed? Can you point me to some examples of live updating within the Bonsai code-base?

I committed my experimental code to the branch: https://github.com/IfcOpenShell/IfcOpenShell/tree/alignment_modeling_experiments

The key alignment building code is https://github.com/IfcOpenShell/IfcOpenShell/blob/0ddd10aff131ebf45cd07f2764b7c1023b4ffac4/src/bonsai/bonsai/tool/alignment.py#L40-L46

**Comment by Andrej730** [comment]
Sure. You can use something simple like `mesh.add_stair` for a reference for general creation of objects:
https://github.com/IfcOpenShell/IfcOpenShell/blob/0ddd10aff131ebf45cd07f2764b7c1023b4ffac4/src/bonsai/bonsai/bim/module/model/stair.py#L131-L172

You'll need to create a new Blender object and `tool.Ifc.link` it to the IFC element (in the stair example it happens during `assign_class` since there is no element yet) and auto-rename it with `tool.Root.set_object_name`.

In case if alignment has a representation you can create shape with `ifcopenshell.geom.create_shape` (or iterator, dependings on how many representations you process at once) create some empty Blender mesh and then apply it to mesh with `tool.Loader.convert_geometry_to_mesh`. Then you can rename mesh using appropriate name from `tool.Loader.get_mesh_name_from_shape` and link ifc representation and mesh with `tool.Geometry.link`.

After everything is done you can run `tool.Collector.assign(obj)` to auto assign the object to some Blender collection.

Not sure what else needed for the start but you can also compare object created by your operator with object created for the similar alignment when project is loaded, all import code is available in `import_ifc.py`.

And highly recomment using VS Code debugger https://github.com/hextantstudios/hextant_python_debugger to debug Blender addon code.

**Reply by RickBrice** [reply]
This is fantastic - thank you so much. It will take a while to understand all of this.

I'm trying to develop the underlying IFCOS alignment API, and building/editing alignments through the Bonsai UI is one of the use cases. I don't think I'll develop a good API if I don't understand the technical mechanisms that support the use case.

**Comment by RickBrice** [comment]
@Andrej730 I'm trying to get the debugger to work. It doesn't stop at breakpoints. The Blender console has this message
<img width="592" alt="image" src="https://github.com/user-attachments/assets/4669df6e-cdb0-41e7-ada7-735ddb808919">

Where would I enter this command?

**Reply by Andrej730** [reply]
> Where would I enter this command?

What command?

Are you sure you setup your VS Code for attaching to debugger server and it attached without isuses? See readme in https://github.com/hextantstudios/hextant_python_debugger

**Comment by RickBrice** [comment]
I think I have everything set up correctly. The only deviation I made is instead of creating a folder for my scripts (e.g.`C:\blender-scripts\addons\my_blender_addon\...`) I just opened `IfcOpenShell\src\bonsai` in VS Code. I've got the live development environment set up so the Bonsai folders in `C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages` have symlinks to `IfcOpenShell\src\bonsai`

Perhaps this is the problem?

When I run a debugging session, and there is an issue with my code, I can see the call stack on both the Blender and VS Code sides. That tells me the Blender and VS Code are attached in the debugging session. 

I have set break points in VS Code, but they will not enable. I can see the red circles before starting debugging, but then they change to open white circles and the debugger doesn't stop at them.

**Reply by Andrej730** [reply]
> Perhaps this is the problem?

Shouldn't be. I have 'IfcOpenShell' repo opened in VS code and extension dependencies are symlinked to the repo and breakpoints work.

> When I run a debugging session, and there is an issue with my code, I can see the call stack on both the Blender and VS Code sides. That tells me the Blender and VS Code are attached in the debugging session.

Okay, if you can see the debug console and callstack, it means connection does work.
You can try to catch some method in the callstack and see the module filepath, perhaps it can give clues about why it's happening (maybe symlinks are not working for some reason in your case). You also can try to set a breakpoint to the file that's not symlinked to be sure this is about symlinks.

> I have set break points in VS Code, but they will not enable. I can see the red circles before starting debugging, but then they change to open white circles and the debugger doesn't stop at them.

It usually happens when you modify the file after it was already loaded in Python. Maybe there are other reasons why VS Code might not match the files.

E.g. if you open Blender so Project Wizard will be visible and put a breakpoint on this line, it doesn't trigger too?

https://github.com/IfcOpenShell/IfcOpenShell/blob/2165dbb3fa38847b60661e5085acf5852bb9a141/src/bonsai/bonsai/bim/module/project/ui.py#L325

**Comment by RickBrice** [comment]
I set a breakpoint as suggested - it does not break.
I did the same using Visual Studio 2022 and Visual Studio Code.

When I open the source folder in VS Code, I tried it from both `C:\IfcOpenShell\src\bonsai` and `C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai`. It doesn't make a difference.

**Reply by Andrej730** [reply]
What about this:

> You can try to catch some method in the callstack and see the module filepath, perhaps it can give clues about why it's happening (maybe symlinks are not working for some reason in your case). You also can try to set a breakpoint to the file that's not symlinked to be sure this is about symlinks.

**Reply by RickBrice** [reply]
Can't get it to break anywhere (code in symlink folders or elsewhere).

This might be a clue, but I don't know. When I open VS Code, before starting debugging, there is the following information
<img width="520" alt="image" src="https://github.com/user-attachments/assets/ec95337f-d46d-4b27-a8a5-b242de15b7d3">

The only place I can get any feedback is if I run my own code, which has a fault in it,
~~~
New cache was created.
Traceback (most recent call last):
  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\tool\ifc.py", line 291, in execute
    IfcStore.execute_ifc_operator(self, context)
  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\ifc.py", line 443, in execute_ifc_operator
    result = getattr(operator, "_execute")(context)
13
 
38
^

  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\alignment\operator.py", line 61, in _execute
    core.build_alignment(tool.Alignment)
  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\core\alignment.py", line 53, in build_alignment
    alignment.build()
  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\tool\alignment.py", line 66, in build
    tool.Collector.assign(mesh)
  File "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\tool\collector.py", line 31, in assign
    for users_collection in obj.users_collection:
28
 
20
^

AttributeError: 'Mesh' object has no attribute 'users_collection'
~~~

**Reply by Andrej730** [reply]
> This might be a clue, but I don't know. When I open VS Code, before starting debugging, there is the following information

It's Pylance warnings, shouldn't be related. You can add extra paths to the project's settings, so it will pick up those folders too.

What if you run in VS Code a script like this - does it works and it breaks?
```python
import debugpy

debugpy.listen(5678)
debugpy.wait_for_client()
# Attatch debugger after this line.

print('hello') # set breakpoint on this line.
```

Also ensure this is disabled in VS Code:
![image](https://github.com/user-attachments/assets/0cbe88e4-2b3a-4483-9da9-34303d06305c)

**Reply by RickBrice** [reply]
I created a test file with your sample script. Without Blender running, I run the test script in VS Code and the breakpoint worked as expected.

Not sure where to check if the Debug Just My Code setting is enabled/disabled in VS Code

**Reply by Andrej730** [reply]
F1 - Settings - "just my code" in search bar

**Reply by RickBrice** [reply]
Toggling that setting did the trick. VS Code now breaks as expected.
Thanks.

Now back to figuring out how bonsai, ifcopenshell, and blender work together.

**Comment by theoryshaw** [comment]
From ai..

Here are some troubleshooting steps that may help:

Check Module Paths: Confirm that the module paths Blender uses align exactly with the paths VS Code sees. You might be able to identify any path mismatches by examining the file paths in the call stack during an active debug session. Even slight differences (like C:\IfcOpenShell\src\bonsai vs. C:\Users\bricer\AppData\Roaming...) could cause VS Code to fail to recognize the files.

Directly Reference Files: Try setting breakpoints directly in the Bonsai files within C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai. This approach bypasses potential symlink issues, which sometimes prevent breakpoints from activating in certain development setups.

Use Explicit Paths in VS Code: Open your Bonsai project in VS Code by navigating to C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai instead of the main project folder. This can ensure the debugger points directly to the exact files Blender is using.

Python Version Compatibility: Verify that the Python version running in Blender matches what VS Code expects. Sometimes mismatches can lead to breakpoints being ignored. Also, check if there are any cached .pyc files that may be causing VS Code to reference outdated bytecode.

Manual Debugging Commands: As a workaround, you can use import pdb; pdb.set_trace() directly in your code where you want the debugger to pause. If this works, it suggests a VS Code or symlink issue rather than a general debugging issue.

Debugger Extensions: Ensure your VS Code Python extension and debugger are updated, as compatibility issues can sometimes affect breakpoint recognition.
